src/rabbit_env.erl

```diff
@@ -1087,8 +1087,34 @@ enabled_plugins_file_from_env(Context) ->
             update_context(Context, enabled_plugins_file, File, environment)
     end.
 
-get_default_enabled_plugins_file(#{config_base_dir := ConfigBaseDir}) ->
-    filename:join(ConfigBaseDir, "enabled_plugins").
+get_default_enabled_plugins_file(#{os_type := {unix, _},
+                                   config_base_dir := ConfigBaseDir,
+                                   data_dir := DataDir}) ->
+    ModernLocation = filename:join(DataDir, "enabled_plugins"),
+    LegacyLocation = filename:join(ConfigBaseDir, "enabled_plugins"),
+    case {filelib:is_regular(ModernLocation),
+          file:read_file_info(LegacyLocation)} of
+        {false, {ok, #file_info{access = read_write}}} ->
+            rabbit_log_prelaunch:info("NOTICE: Using 'enabled_plugins' file"
+                                      " from ~p. Please migrate this file"
+                                      " to its new location, ~p, as the"
+                                      " previous location is deprecated.",
+                                      [LegacyLocation, ModernLocation]),
+            LegacyLocation;
+        {false, {ok, #file_info{access = read}}} ->
+            {ok, _} = file:copy(LegacyLocation, ModernLocation),
```

Comment:
This module can't have side effects like creating files. Its purpose is only to interpret the environment and return a computed context.

In fact this entire logic should live in the `rabbit_plugins` module in rabbitmq-server, which is responsible for the `enabled_plugins` file.

The context can store the old and new locations to communicate them to `rabbit_plugins`.

Reply:
This generally did cross my mind, but I didn't think it was worth the greater complexity of changes. I willing to be convinced, though.

So, I think the cleanest way to do that is to move the notion of the default all the way out of the context, otherwise the context has to communicate either:
- a location that should be used unequivocally, because it came directly for the env var
- two locations that should be handled exactly with the described fallback behavior
The context could return somelike `{certainly, "some user specified location"}` or `{with_fallback, ["/var/lib/rabbitmq/enabled_plugins", "/etc/rabbitmq/enabled_plugins"]}`, but the coupling between the modules/repos doesn't feel right to me.

Reply:
@dumbbell Is there a reasonable possibility of the context returning the location as a kind of lazily executed, but memoized deferred value, that the caller can resolve at the time of its choosing (I'm thinking of https://clojuredocs.org/clojure.core/delay)? Then the broker and cli would only need to trigger resolution, instead of both having a the fallback mechanism. I suspect the answer is no, since gen_server seems to fulfill this kind of role generally in erlang.